src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -0,0 +1,288 @@
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gcs_actor_manager.h"
+#include <ray/common/ray_config.h>
+#include <ray/gcs/redis_gcs_client.h>
+
+#include <utility>
+
+namespace ray {
+namespace gcs {
+
+ClientID GcsActor::GetNodeID() const {
+  const auto &raylet_id_binary = actor_table_data_.address().raylet_id();
+  if (raylet_id_binary.empty()) {
+    return ClientID::Nil();
+  }
+  return ClientID::FromBinary(raylet_id_binary);
+}
+
+void GcsActor::UpdateAddress(const rpc::Address &address) {
+  actor_table_data_.mutable_address()->CopyFrom(address);
+}
+
+const rpc::Address &GcsActor::GetAddress() const { return actor_table_data_.address(); }
+
+WorkerID GcsActor::GetWorkerID() const {
+  const auto &address = actor_table_data_.address();
+  if (address.worker_id().empty()) {
+    return WorkerID::Nil();
+  }
+  return WorkerID::FromBinary(address.worker_id());
+}
+
+void GcsActor::UpdateState(rpc::ActorTableData::ActorState state) {
+  actor_table_data_.set_state(state);
+}
+
+rpc::ActorTableData::ActorState GcsActor::GetState() const {
+  return actor_table_data_.state();
+}
+
+ActorID GcsActor::GetActorID() const {
+  return ActorID::FromBinary(actor_table_data_.actor_id());
+}
+
+TaskSpecification GcsActor::GetCreationTaskSpecification() const {
+  const auto &task_spec = actor_table_data_.task_spec();
+  return TaskSpecification(task_spec);
+}
+
+const rpc::ActorTableData &GcsActor::GetActorTableData() const {
+  return actor_table_data_;
+}
+
+rpc::ActorTableData *GcsActor::GetMutableActorTableData() { return &actor_table_data_; }
+
+/////////////////////////////////////////////////////////////////////////////////////////
+GcsActorManager::GcsActorManager(
+    std::shared_ptr<gcs::RedisGcsClient> redis_gcs_client,
+    gcs::GcsNodeManager &gcs_node_manager,
+    std::unique_ptr<gcs::GcsActorScheduler> gcs_actor_scheduler)
+    : redis_gcs_client_(redis_gcs_client),
+      gcs_actor_scheduler_(std::move(gcs_actor_scheduler)) {
+  RAY_LOG(INFO) << "Initializing GcsActorManager.";
+  gcs_node_manager.AddNodeAddedListener(
+      [this](const std::shared_ptr<rpc::GcsNodeInfo> &) {
+        // Because a new node has been added, we need to try to schedule the pending
+        // actors.
+        SchedulePendingActors();
+      });
+
+  gcs_node_manager.AddNodeRemovedListener([this](std::shared_ptr<rpc::GcsNodeInfo> node) {
+    // All of the related actors should be reconstructed when a node is remove from gcs
+    // node manager.
+    ReconstructActorsOnNode(ClientID::FromBinary(node->node_id()));
+  });
+
+  gcs_actor_scheduler_->SetScheduleFailureHandler(
```

Comment:
> Is there a reason why these are added dynamically instead of in the constructor?

Well, there's no specific reason for this, but I've moved it to the constructor.